### PR TITLE
Mekhanik evgenii/fix 525 3

### PR DESCRIPTION
**Pull request by EvgeniiMekhanik**

### CODE REVIEW

File: fw/tls.c [hunk 1]
```diff
@@ -100,19 +101,24 @@ tfw_tls_connection_recv(TfwConn *conn, struct sk_buff *skb)
 		BUG_ON(r == T_BLOCK);
 		fallthrough;
 	case T_BAD:
-		r = T_BAD;
 		if (tls->conf->endpoint == TTLS_IS_SERVER)
 			TFW_INC_STAT_BH(serv.tls_hs_failed);
 		fallthrough;
 	case T_BLOCK_WITH_FIN:
+		r = T_BLOCK_WITH_FIN;
```
Comment (const-t) [inline fw/tls.c:108]: Why do we assign `T_BLOCK_WITH_FIN` to `r` here?

File: fw/tls.c [hunk 2]
```diff
@@ -174,11 +180,23 @@ tfw_tls_connection_recv(TfwConn *conn, struct sk_buff *skb)
...
+		else
+			r = save_err_code;
+
+		if (r == T_BLOCK_WITH_FIN || r == T_BLOCK_WITH_RST) {
 			kfree_skb(nskb);
 			return r;
+		} else if (r == T_BAD) {
+			/*
+			 * In case of T_BAD error we close connection
+			 * with tcp_shgutdown() and gracefully send all
```
Comment (const-t) [inline fw/tls.c:194]: typo tcp_shutdown

File: fw/tls.c [hunk 3]
```diff
@@ -174,11 +180,23 @@ tfw_tls_connection_recv(TfwConn *conn, struct sk_buff *skb)
...
+			 * pending responses to client. We should continue
+			 * to process WINDOW_UPDATE frames so, we should
+			 * decrypt all skbs, not drop them.
+			 */
+			save_err_code = r;
```
Comment (const-t) [inline fw/tls.c:199]: `save_err_code` is really needed here?

**Review by const-t — APPROVED** [review]

File: fw/sock.c [hunk 4]
```diff
@@ -805,7 +820,12 @@ ss_tcp_process_skb(struct sock *sk, struct sk_buff *skb, int *processed)
 		 */
 		BUG_ON(conn == NULL);
 
-		if (SS_CONN_TYPE(sk) & Conn_Stop) {
+		/*
+		 * We should continue to proess WINDOW_UPDATE frames
+		 * for closing connections.
+		 */
+		if ((SS_CONN_TYPE(sk) & Conn_Stop)
+		    && !((SS_CONN_TYPE(sk) & Conn_H2Clnt) == Conn_H2Clnt)) {
```
Comment (krizhanovsky) [inline fw/sock.c:828]: It's quite bad that the low level socket API must know about HTTP/2, which is 2 layers higher. Is this absolutely necessary?

File: fw/sock.c [hunk 5]
```diff
@@ -1414,14 +1456,19 @@ ss_tx_action(void)
 
 		bh_lock_sock(sk);
 		if (sock_flag(sk, SOCK_DEAD)) {
```
Comment (krizhanovsky) [inline fw/sock.c:1458]: What about this comment https://github.com/tempesta-tech/tempesta/pull/2007/files#r1406949163 ? At least a code comment must be added

**Review by krizhanovsky — COMMENTED** [review]
I didn't find any sever issues with the PR except removing a logical modules separation, which worsen the architecture. Let's discuss this on the call.

File: fw/tls.c [hunk 6]
```diff
@@ -166,10 +180,23 @@ tfw_tls_connection_recv(TfwConn *conn, struct sk_buff *skb)
 		spin_unlock(&tls->lock);
 
 		/* Do upcall to http or websocket */
-		r = tfw_connection_recv(conn, data_up.skb);
-		if (r == T_BLOCK || r == T_BAD) {
+		if (save_err_code == T_OK)
```
[the remark this answers is not in the document]
Comment (EvgeniiMekhanik) [reply fw/tls.c:183]: We discuss it on the meeting.  If `tfw_connection_recv` fails with T_BAD we save this value in  `save_err_code` variable, then we decrypt all pending skbs and return save_err_code.

File: fw/connection.c [hunk 7]
```diff
@@ -69,13 +69,21 @@ tfw_connection_repair(TfwConn *conn)
 int
 tfw_connection_close(TfwConn *conn, bool sync)
 {
-	return TFW_CONN_HOOK_CALL(conn, conn_close, sync);
+	int r;
+
+	tfw_connection_get(conn);
```
[the remark this answers is not in the document]
Comment (EvgeniiMekhanik) [reply fw/connection.c:86]: Yes fixed

File: fw/tls.c [hunk 8]
```diff
@@ -542,6 +547,29 @@ tfw_tls_send(TlsCtx *tls, struct sg_table *sgt)
 	if (ttls_xfrm_need_encrypt(tls))
 		flags |= SS_SKB_TYPE2F(io->msgtype) | SS_F_ENCRYPT;
 
+	if (io->msgtype == TTLS_MSG_ALERT &&
+	    (io->alert[1] == TTLS_ALERT_MSG_CLOSE_NOTIFY ||
+	     io->alert[0] == TTLS_ALERT_LEVEL_FATAL)) {
+		switch (io->st_flags
+			& (TTLS_F_ST_SHUTDOWN | TTLS_F_ST_CLOSE)) {
```
[the remark this answers is not in the document]
Comment (EvgeniiMekhanik) [reply fw/tls.c]: Yes, fixed

File: fw/sock.c (hunk 5 above)
Comment (EvgeniiMekhanik) [reply fw/sock.c:1458]: I add comment about

File: fw/sock.c (hunk 4 above)
Comment (EvgeniiMekhanik) [reply fw/sock.c:828]: Fixed

File: fw/sock.c [hunk 9]
```diff
@@ -38,10 +38,12 @@
 #include "tempesta_fw.h"
 #include "work_queue.h"
 #include "http_limits.h"
+#include "connection.h"
```
[the remark this answers is not in the document]
Comment (EvgeniiMekhanik) [reply fw/sock.c:41]: Fixed

File: fw/tls.c (hunk 1 above)
Comment (EvgeniiMekhanik) [reply fw/tls.c]: Yes fixed

**Review by krizhanovsky — APPROVED** [review]
LGTM. Thank you for the fixes!

**Pull request merged**